py/objtuple.c

```diff
@@ -31,6 +31,10 @@
 #include "py/objtuple.h"
 #include "py/runtime.h"
 
+// type check is done on getiter method to allow tuple, namedtuple, attrtuple
+// Assumes mp_obj_is_obj has already been checked
+#define _mp_obj_is_tuple_compatible(o) ((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type->getiter == mp_obj_tuple_getiter
```

Comment:
It may not be that the object is a true object.  Eg it could be called as `mp_obj_tuple_get(MP_OBJ_NEW_SMALL_INT(1), ...)`.  Of course this is silly but the existing code will catch this (if assertions are enabled), which is exactly what the assertion is for.